fem/libceed/coefficient.hpp

```diff
@@ -0,0 +1,216 @@
+   {
+      CeedVectorDestroy(&coeffVector);
+   }
+};
+
+struct GridCoefficient : VariableCoefficient
+{
+   const mfem::GridFunction &gf;
+   CeedBasis basis;
+   CeedElemRestriction restr;
+   GridCoefficient(const mfem::GridFunction &gf_)
+      : VariableCoefficient(gf_.VectorDim(), CEED_EVAL_INTERP),
+        gf(gf_)
+   {
+      InitVector(gf, coeffVector);
+   }
+};
+
+struct QuadCoefficient : VariableCoefficient
+{
+   mfem::Vector coeff;
+   CeedElemRestriction restr;
+   QuadCoefficient(int ncomp_) : VariableCoefficient(ncomp_, CEED_EVAL_NONE) { }
+};
+
+/** @brief Identifies the type of coefficient of the Integrator to initialize
+    accordingly the CeedData. */
+template <typename Context>
+void InitCoefficient(mfem::Coefficient *Q, mfem::Mesh &mesh,
+                     const mfem::IntegrationRule &ir,
+                     Coefficient*& coeff_ptr, Context &ctx)
+{
+   if ( Q == nullptr )
+   {
+      Coefficient *ceedCoeff = new Coefficient(1);
+      ctx.coeff = 1.0;
+      coeff_ptr = ceedCoeff;
+   }
+   else if (ConstantCoefficient *coeff = dynamic_cast<ConstantCoefficient*>(Q))
+   {
+      Coefficient *ceedCoeff = new Coefficient(1);
+      ctx.coeff = coeff->constant;
+      coeff_ptr = ceedCoeff;
+   }
+   else if (GridFunctionCoefficient* coeff =
+               dynamic_cast<GridFunctionCoefficient*>(Q))
+   {
+      GridCoefficient *ceedCoeff = new GridCoefficient(*coeff->GetGridFunction());
+      coeff_ptr = ceedCoeff;
+   }
+   else if (QuadratureFunctionCoefficient *cQ =
+               dynamic_cast<QuadratureFunctionCoefficient*>(Q))
+   {
+      QuadCoefficient *ceedCoeff = new QuadCoefficient(1);
+      const int ne = mesh.GetNE();
+      const int nq = ir.GetNPoints();
+      const mfem::QuadratureFunction &qFun = cQ->GetQuadFunction();
+      MFEM_VERIFY(qFun.Size() == nq * ne,
+                  "Incompatible QuadratureFunction dimension \n");
+
+      MFEM_VERIFY(&ir == &qFun.GetSpace()->GetElementIntRule(0),
+                  "IntegrationRule used within integrator and in"
+                  " QuadratureFunction appear to be different");
+      qFun.Read();
+      ceedCoeff->coeff.MakeRef(const_cast<mfem::QuadratureFunction &>(qFun),0);
+      InitVector(ceedCoeff->coeff, ceedCoeff->coeffVector);
+      coeff_ptr = ceedCoeff;
+   }
+   else
+   {
+      QuadCoefficient *ceedCoeff = new QuadCoefficient(1);
+      const int ne = mesh.GetNE();
+      const int nq = ir.GetNPoints();
+      ceedCoeff->coeff.SetSize(nq * ne);
+      auto C = Reshape(ceedCoeff->coeff.HostWrite(), nq, ne);
+      for (int e = 0; e < ne; ++e)
+      {
+         mfem::ElementTransformation &T = *mesh.GetElementTransformation(e);
+         for (int q = 0; q < nq; ++q)
+         {
+            C(q,e) = Q->Eval(T, ir.IntPoint(q));
+         }
+      }
+      InitVector(ceedCoeff->coeff, ceedCoeff->coeffVector);
+      coeff_ptr = ceedCoeff;
+   }
+}
+
+/** @brief Identifies the type of vector coefficient of the Integrator to
```

Comment:
Same comment as above.

Reply:
I improved the documentation, tell me if you like it.